crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,313 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use crate::bad_token::{BadTokenDetecting, TokenQuality};
+use crate::price_estimation::gas::GAS_PER_WETH_UNWRAP;
+use anyhow::Result;
+use model::order::BUY_ETH_ADDRESS;
+use primitive_types::H160;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+/// Verifies that buy and sell tokens are supported and handles
+/// ETH as buy token appropriately.
+pub struct SanitizedPriceEstimator<T: PriceEstimating> {
+    inner: T,
+    bad_token_detector: Arc<dyn BadTokenDetecting>,
+    native_token: H160,
+}
+
+type EstimationResult = Result<Estimate, PriceEstimationError>;
+
+enum EstimationProgress {
+    TrivialSolution(EstimationResult),
+    AwaitingEthEstimation,
+    AwaitingErc20Estimation,
+}
+
+impl<T: PriceEstimating> SanitizedPriceEstimator<T> {
+    pub fn new(
+        inner: T,
+        native_token: H160,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+    ) -> Self {
+        Self {
+            inner,
+            native_token,
+            bad_token_detector,
+        }
+    }
+
+    async fn get_token_quality_errors(
+        &self,
+        queries: &[Query],
+    ) -> HashMap<H160, PriceEstimationError> {
+        let mut token_quality_errors: HashMap<H160, PriceEstimationError> = Default::default();
+        let mut checked_tokens = HashSet::<H160>::default();
+
+        for token in queries
+            .iter()
+            .copied()
+            .flat_map(|query| [query.buy_token, query.sell_token])
+        {
+            if checked_tokens.contains(&token) {
+                continue;
+            }
+
+            match self.bad_token_detector.detect(token).await {
+                Err(err) => {
+                    token_quality_errors.insert(
+                        token,
+                        PriceEstimationError::Other(crate::clone_anyhow_error(&err)),
```

Comment:
Can't we just move the `err` here instead of cloning?
```suggestion
                        PriceEstimationError::Other(err),
```

Reply:
Moving is not possible here.
Because we store the errors in a map, we only have at most 1 PriceEstimationError per token. If we have 5 Queries, all using the same unsupported token, we need to clone this single error into 5 different results.

Reply:
So, I agree we need to clone the `PriceEstimationError` when checking that the tokens are supported here (as you said, this error may be for more than one query):
```rust
                if let Some(err) = token_quality_errors.get(&query.buy_token) {
                    return EstimationProgress::TrivialSolution(Err(err.clone()));
                }
                if let Some(err) = token_quality_errors.get(&query.sell_token) {
                    return EstimationProgress::TrivialSolution(Err(err.clone()));
                }
```

But I'm not 100% sure I understand why we can't move the error into the `HashMap` (since we only insert a single error per token and the `err` is dropped at the end of the `match` branch anyway).

That being said, if the compiler says "no" then it knows better 😛.

Reply:
Ah, sorry. Greped for `clone_anyhow_error` and found the wrong one. 🤦‍♂️ 
When the error gets put into the map it can be moved, indeed.
Good catch!